### PR TITLE
ddl: reorganize in conversion between double and float

### LINKED ISSUE #31372: double column  value are different with MySQL after changing column type from float to double
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```sql
create table ft(id int primary key auto_increment, tmp float(6,2));
insert into ft(tmp) values(36.4);
insert into ft(tmp) values(36.4);
insert into ft(tmp) values(36.4);
select * from ft;
alter table ft modify tmp double(6,2);
select * from ft;
```

### 2. What did you expect to see? (Required)
MySQL - 5.7.22
```
mysql> create table ft(id int primary key auto_increment, tmp float(6,2));
Query OK, 0 rows affected (0.03 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.01 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.00 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.00 sec)

mysql> select * from ft;
+----+-------+
| id | tmp   |
+----+-------+
|  1 | 36.40 |
|  2 | 36.40 |
|  3 | 36.40 |
+----+-------+
3 rows in set (0.01 sec)

mysql> alter table ft modify tmp double(6,2);
Query OK, 3 rows affected (0.05 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select * from ft;
+----+-------+
| id | tmp   |
+----+-------+
|  1 | 36.40 |
|  2 | 36.40 |
|  3 | 36.40 |
+----+-------+
3 rows in set (0.00 sec)

mysql> select version();
+------------+
| version()  |
+------------+
| 5.7.22-log |
+------------+
1 row in set (0.01 sec)
```

### 3. What did you see instead (Required)
TiDB-5.3.0
```
mysql> create table ft(id int primary key auto_increment, tmp float(6,2));
Query OK, 0 rows affected (0.07 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.01 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.01 sec)

mysql> insert into ft(tmp) values(36.4);
Query OK, 1 row affected (0.00 sec)

mysql>  select * from ft;
+----+------+
| id | tmp  |
+----+------+
|  1 | 36.4 |
|  2 | 36.4 |
|  3 | 36.4 |
+----+------+
3 rows in set (0.00 sec)

mysql> alter table ft modify tmp double(6,2);
Query OK, 0 rows affected (0.07 sec)

mysql>  select * from ft;
+----+--------------------+
| id | tmp                |
+----+--------------------+
|  1 | 36.400001525878906 |
|  2 | 36.400001525878906 |
|  3 | 36.400001525878906 |
+----+--------------------+
3 rows in set (0.00 sec)

mysql> select tidb_version()
    -> ;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.3.0
Edition: Community
Git Commit Hash: 4a1b2e9fe5b5afb1068c56de47adb07098d768d6
Git Branch: heads/refs/tags/v5.3.0
UTC Build Time: 2021-11-24 13:31:24
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)
```
mysql> select tidb_version()
    -> ;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.3.0
Edition: Community
Git Commit Hash: 4a1b2e9fe5b5afb1068c56de47adb07098d768d6
Git Branch: heads/refs/tags/v5.3.0
UTC Build Time: 2021-11-24 13:31:24
GoVersion: go1.16.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: close #31372 

Problem Summary:

### What is changed and how it works?
According to https://github.com/pingcap/tidb/issues/31372#issuecomment-1006382531, change data is necessary for conversion between float and double.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix double column value are different with MySQL after changing column type from float to double
```

### CODE REVIEW

**Review by Defined2014 — APPROVED** [review]

**Comment by unconsolable** [comment]
/cc @AilinKid 
PTAL, thanks.

**Review by AilinKid — APPROVED** [review]
/LGTM

**Comment by AilinKid** [comment]
> /cc @AilinKid PTAL, thanks.

Great job! (｡･ω･｡)

File: ddl/column_type_change_test.go [hunk 1]
```diff
@@ -2352,3 +2352,19 @@ func (s *testColumnTypeChangeSuite) TestChangeNullValueFromOtherTypeToTimestamp(
...
+	tk.MustExec("use test")
+
+	prepare := func() {
+		tk.MustExec("drop table if exists t;")
+		tk.MustExec("create table t (a float(6,2));")
+		tk.MustExec("insert into t values (36.4), (24.1);")
+	}
+
+	prepare()
+	tk.MustExec("alter table t modify a double(6,2)")
```
Comment (zimulala) [inline ddl/column_type_change_test.go:2368]: Please add some tests like double(6,2) -> double(6,1) and double -> float.

Comment (unconsolable) [reply ddl/column_type_change_test.go:2368]: Added.

**Review by bb7133 — APPROVED** [review]

**Pull request merged**